tools/outline_proxy_controller/outline_proxy_controller.cpp

```diff
@@ -29,6 +34,93 @@ using namespace outline;
 
 extern Logger logger;
 
+/**
+ * @brief
+ * Create a new child process of `cmd` with arguments `args`, and return its
+ * process id as well as the redirected stdout/stderr stream.
+ *
+ * This function is similar to popen, but popen executes an arbitrary command
+ * in a shell context which opens a hole of os command injection; while this
+ * function will make sure we only execute `cmd` itself.
+ * 
+ * @param cmd The process name to be executed.
+ * @param args The arguments of the process (don't include `cmd` itself).
+ * @return std::tuple<pid_t, FILE*> The process ID and its stdout/stderr stream.
+ */
+static tuple<pid_t, FILE*> safe_popen(const string &cmd, const vector<string> &args) {
+  // pipefd[0] is read-only; pipefd[1] is write-only
+  int pipefd[2];
+  if (pipe(pipefd) == -1) {
+    throw runtime_error("failed to create pipe for " + cmd + " command");
+  }
+
+  pid_t pid;
+  switch (pid = fork()) {
+  case -1:
+    close(pipefd[0]);
+    close(pipefd[1]);
+    throw runtime_error("failed to fork a new process for " + cmd + " command");
+  case 0:
+    // child process:
+    //   redirect stdout/stderr to write pipe and close read pipe
+    close(pipefd[0]);
+    dup2(pipefd[1], STDOUT_FILENO);
+    dup2(pipefd[1], STDERR_FILENO);
```

Comment:
Doesn't popen only pipe the stdout? Maybe we shouldn't capture the stderr, and just output it as normal.

Reply:
Actually, I wonder if capturing the STDERR will break things. We should probably preserve the behavior.

Reply:
Yes `popen` only redirects stdout, but our code redirects `stderr` to `stdout`:

* https://github.com/Jigsaw-Code/outline-client/blob/master/tools/outline_proxy_controller/outline_proxy_controller.cpp#L73
* https://github.com/Jigsaw-Code/outline-client/blob/master/tools/outline_proxy_controller/outline_proxy_controller.h#L178